Development

mcrd25 opened the pull request:
We did this using TDD approach.

We are aware that the next project requires the same with RSpec, can we use the same pull request for the next project once you have reviewed this one? Let us know

Commit:
Initialized RSpec

Commit:
Merge pull request #1 from adriaanbd/initial-setup

Initialized RSpec

Commit:
Started cell class and test for same

adriaanbd committed:
Completed Cell class with test

adriaanbd committed:
Merge pull request #2 from adriaanbd/cell

Cell

adriaanbd committed:
Initial player class with test

Commit:
Added test to player_spec

Commit:
Merge pull request #3 from adriaanbd/player

Player

Commit:
Created files for Board class and set up initial test

adriaanbd committed:
Added custom grid attribute and default grid method with test

adriaanbd committed:
Adding #get_cell with spec

adriaanbd committed:
Add #set_cell with spec

adriaanbd committed:
#game_over w spec and winner? draw? stubbed not yet implemented

Commit:
Added core extensions file with extensions on array class to be used next in board class

adriaanbd committed:
Added #winner? and #draw? with spec

Commit:
Merge pull request #4 from adriaanbd/board

Board

Commit:
Created game class and game_spec file and created tests and methods for initialize and switch_players respectively

adriaanbd committed:
Added #solicit_move, #get_move, #game_over_mssg and #keyboard_to_position with spec

Commit:
Completed game methods and added formatted_grid function to board class for display in command line

Commit:
Additional changes

Commit:
Merge pull request #5 from adriaanbd/game

Game

Commit:
Updated formatted_grid in board class to display number on grid when empty

adriaanbd committed:
Updating #formatted grid; not done.

Commit:
Finalised change in formatted_grid function

Commit:
Started test for formatted grid function

adriaanbd committed:
Bucketed Formatted Grid test and updated #play

Commit:
Added main script for game

Commit:
Merge pull request #6 from adriaanbd/upgrades

Upgrades

heyapricot commented on lib/classes/game.rb:
```diff
@@ -0,0 +1,70 @@
+			while true
+				board.formatted_grid
+    		puts ''
+    		puts solicit_move
```
- This could be a matter of style, but I don't like the `puts ' '`
- For `solicit_move` can we put the print inside the method?
- Same with `board.formatted_grid`, can we wrap it around a method? `print_board` or something similar

mcrd25 replied:
Hi, 

- okay, would use `\n`
- Would print statement within `solicit_move` method instead. 
- Not sure I understand your third statement. Do you want us to wrap elements of the `formatted_grid` method in another method or put `board.formatted_grid` inside a method `print_board` for stylistic reasons?

mcrd25 replied:
Resolved with commits [1133cdeac5f1af982b7c860d5af754e30af70791](https://github.com/adriaanbd/tic-tac-toe/tree/1133cdeac5f1af982b7c860d5af754e30af70791) and [00d9cefeb174b304924e85928ad39cdf5dba886e](https://github.com/adriaanbd/tic-tac-toe/tree/00d9cefeb174b304924e85928ad39cdf5dba886e)

heyapricot commented on lib/classes/game.rb:
```diff
@@ -0,0 +1,70 @@
+				'9' => [2, 2]
+			}
+			keyboard_map[move]
+		end
```
Cool idea 😎

heyapricot replied:
Maybe I'm just lazy, but this seems like a lot of writing (Also, it makes us vulnerable to a typo). What if we had a method that populated the hash? The hash keys are just a sequence from `(1..9)`. Can we use math to generate the array to which the index maps? 🤔🧮

mcrd25 replied:
Hi, thanks! You are right it is possibly best to write a method to populate the hash.
Would check it out now. Do we HAVE to do this now or can we do this at a later date?

mcrd25 replied:
Issue #9 was created to address this suggested improvement at a later date when we have more time

heyapricot replied:
It can be done later. Good idea saving it as an issue

heyapricot commented on lib/classes/game.rb:
```diff
@@ -0,0 +1,70 @@
+    		puts solicit_move
+				col, row = get_move
+				puts ''
+    		move = board.set_cell(row, col, @current_player.turn)
```
Why is the player storing the current turn? If we have two players that means we are storing the turn twice, right? There should be a better way...

mcrd25 replied:
Hi sorry, misunderstanding perhaps we need to rename this variable turn is basically if you are X or O

mcrd25 replied:
Resolved with commit [00d9cefeb174b304924e85928ad39cdf5dba886e](https://github.com/adriaanbd/tic-tac-toe/pull/7/commits/b873d310d67dbfea35ea1959259e856ce3c9c567)

heyapricot commented on lib/classes/board.rb:
```diff
@@ -0,0 +1,77 @@
+
+		def set_cell(row, col, value)
+			cell = get_cell(row, col)
+			cell.value.empty? ? cell.value = value : nil
```
I like the [`||=`](http://www.rubyinside.com/what-rubys-double-pipe-or-equals-really-does-5488.html) operator for this kind of thing

adriaanbd replied:
After considering this suggestion we decided not to go with it for the following reasons:
1. We don't want to overwrite the cell if it has a value
2. We need this method to return nil if its not empty because our #play method requires it.
```
move = board.set_cell(row, col, @current_player.turn)
while move.nil?
   puts solicit_move + ' Please enter a number in an empty cell.'
   col, row = get_move
   puts ''
   move = board.set_cell(row, col, @current_player.turn)
end
```

heyapricot replied:
I'll explain myself better.
1) First of all, for cell.value I don't think you need to store the empty string. Why not initialize cell value to `nil`? 
2) If the previous change was implemented you could do something like
```
    def set_cell(row, col, value)
      cell = get_cell(row, col)
      # Write cell.value only if it's nil
      cell.value ||= value
      # Return if the cell value is the same as the one you got as argument. This method will return true if the wirte was succesful and false otherwise
      cell.value == value
      # At this point i'm not sure why do you need this method to return nil
    end
```

heyapricot replied:
> We don't want to overwrite the cell if it has a value

That's exactly what the `||=` operator does

mcrd25 replied:
Thanks, I actually really like what we did.

tried ||= does not seem to set the value, I don't know why. hmmm....

mcrd25 replied:
Oh right, because the initial value of the cell is ""

mcrd25 replied:
I will initialize the cell value to nil as you suggested. It would make some breaking changes but I guess we can do that to accommodate your suggestion! thanks :)

mcrd25 replied:
Wow, this change is causing problems, for some reason, the function does not return anything other than what is returned at ||= statement event when we explicitly say return cell.value == value it still returns the value or nil. Weird. I think we are missing something but perhaps we can make an issue out of this and move on since it does not affect functionality and is a matter of personal preference.

We will leave it as it was before.

See iss #17

mcrd25 replied:
See commit [4d1f2abce5779b64ea7bf3ad9d38d989c06aafd9](https://github.com/adriaanbd/tic-tac-toe/pull/18/commits/4d1f2abce5779b64ea7bf3ad9d38d989c06aafd9) to see that we attempted.

We reverted back to the original version in commit [a90931c9cbcf66e3b33aa777b90d3557fee8fbf1](https://github.com/adriaanbd/tic-tac-toe/pull/18/commits/a90931c9cbcf66e3b33aa777b90d3557fee8fbf1)

heyapricot commented on lib/classes/board.rb:
```diff
@@ -0,0 +1,77 @@
+	class Board
+		attr_reader :grid
+		def initialize(grid_board = {})
+			@grid = grid_board.fetch(:grid, default_board)
```
What's the point on wrapping the board array in a hash? In the end you are still storing the moves in an array (Which is inside a hash)

adriaanbd replied:
We see three advantages to this:

1. We want to be able to initialize a board with a custom grid, as we've done in the board_spec's `#get_cell` for example.

2. We want to return a default grid whenever we do not provide a custom grid.

3. Although we can stub the grid for the test and leave only the default grid as a default argument for initialization, we'd like to leave open the possibility for initialization with a custom grid.

heyapricot replied:
I don't think any of those reasons justify having the grid array inside a hash. 
1. You can define a method to write the state of the board or make the grid property read/write with :attr_accessor (Which is not very safe, I'd go with the first option).
2. You can have your default grid withoud using the hash
3. Same as 2 and 1

My point is: For the tic-tac-toe logic, the project is using an array. The array is the one that's storing all the elements in the tic-tac-toe. In the current implementation there is an extra step where I have to retrieve the array from the hash and then manipulate the array. You can take the hash away and manipulate the array directly. I truly see no reason in the code to have the array inside a hash.

mcrd25 replied:
Thanks, we see what you mean.
Resolved with commit [a90931c9cbcf66e3b33aa777b90d3557fee8fbf1](https://github.com/adriaanbd/tic-tac-toe/pull/18/commits/a90931c9cbcf66e3b33aa777b90d3557fee8fbf1)

heyapricot commented on lib/classes/cell.rb:
```diff
@@ -0,0 +1,8 @@
+		def initialize(value = '')
+			@value = value
+		end
+	end
```
🤔 If we are only using the `Cell` class to store a value, do we really need a class?

mcrd25 replied:
Not right now but being true to OOP I think it useful especially if we are to at a later day change the behavior of a cell. For example, change the color of a cell if the cell is part of a winning position.

heyapricot commented on lib/classes/game.rb:
```diff
@@ -0,0 +1,70 @@
+					col, row = get_move
+					puts ''
+    			move = board.set_cell(row, col, @current_player.turn)
+    		end	
```
Please watch the indentation of your files.
We can move the while block inside a method. It would look nicer

mcrd25 replied:
Yes, thank you.

Resolved with commit [b873d310d67dbfea35ea1959259e856ce3c9c567](https://github.com/adriaanbd/tic-tac-toe/pull/7/commits/b873d310d67dbfea35ea1959259e856ce3c9c567)

heyapricot commented on lib/classes/game.rb:
```diff
@@ -0,0 +1,70 @@
+					return
+				else
+					switch_players
+				end
```
If we move `switch_players` to another part of the while loop we can rewrite the previous block to something like:
`break if @board.game_over`
We can then call game_over_mssg out of the while loop. What do you think?

mcrd25 replied:
Thanks for observation
This issue resolved with commit [dcc5b42c1ac66ed0ed8d636adf516286fbe18047](https://github.com/adriaanbd/tic-tac-toe/pull/7/commits/dcc5b42c1ac66ed0ed8d636adf516286fbe18047)

heyapricot commented on lib/classes/board.rb:
```diff
@@ -0,0 +1,77 @@
+							inner + 4
+						when 2
+							inner + 7
+						end
```
If we use math, we can get rid of the `case` block. 
The three operations are very similar: `inner + number`
Which statement would allow us to get `1,4,7`? Use the elements you have: `outer`, `inner`

mcrd25 replied:
We can use math that's how we originally got those numbers but we would still need conditionals.

mcrd25 replied:
Thanks changed to `inner +  1 + (3 * outer)` and fixed in commit [58ce57a874c05da88f649f70797996c8c6e14e78](https://github.com/adriaanbd/tic-tac-toe/pull/7/commits/58ce57a874c05da88f649f70797996c8c6e14e78)

heyapricot commented on lib/classes/board.rb:
```diff
@@ -0,0 +1,77 @@
+
+		def winning_positions
+			grid + grid.transpose + diagonals
+		end
```
Not bad *per se* but we are generating a whole new matrix by calling `grid.transpose` just to make the checks. You can leave it like this, but I want you to think about performance when building apps. Do we really need to generate a whole new matrix just to check for the columns?

Life is full of tradeoffs. In programming it may come in the form of commodity vs performance.

mcrd25 replied:
Okay great, thank.

We would create an issue out of this and see if we can address it at a later date. 

Thanks again. See iss #13

heyapricot commented on lib/classes/board.rb:
```diff
@@ -0,0 +1,77 @@
+				[get_cell(0, 0), get_cell(1, 1), get_cell(2, 2)],
+				[get_cell(0, 2), get_cell(1, 1), get_cell(2, 0)]
+			]
+		end
```
I try to avoid hardcoded values when possible. Can we generate the combinations for the diagonals?

mcrd25 replied:
Created an issue #15  out of it to solve at later date. We really appreciate the observation!

heyapricot commented on lib/classes/core_extensions.rb:
```diff
@@ -0,0 +1,17 @@
+	def none_empty?
+		!any_empty?
+	end
+end
```
This is a clever idea but I'm not crazy about the extension of core elements. This discussion enters the realm of personal preference/philosophy. [This is the best info I could find about it ](https://stackoverflow.com/questions/16504369/is-it-proper-ruby-style-to-extend-built-in-classes)

mcrd25 replied:
Hi, we like it. It also allows us to implement some of the skills we learned in the previous project.

But we really appreciate your comment and perhaps we can discuss at a later date this issue.

adriaanbd replied:
> This is a clever idea but I'm not crazy about the extension of core elements. This discussion enters the realm of personal preference/philosophy. [This is the best info I could find about it ](https://stackoverflow.com/questions/16504369/is-it-proper-ruby-style-to-extend-built-in-classes)

Thanks for the informative link. I echo Ben Miller's comment:
"[...] Monkey patching would be redefining the functionality of an existing method on the base class with out creating a new type. Basically something like making Array.each() do something different. What you are doing here is not polymorphism, it is inheritance and is the back bone of OOP."

heyapricot reviewed (changes requested):
# Overkill 💣
Hey! Congratulations on building your first big ruby project! The game works and you even added some tests, which is very nice. However, there are elements that I think are overkill. They could be simplified. I've left some comments. Please address them and after send a new code review request.

- Additional observation: I like the idea of using a module but it's not implemented appropiately. `Board` is the only class inside the module. Either avoid the use of the module for now or group all the classes into a single module

As always feel free to reach me on crisp for any kind of help/support request. Thanks

mcrd25 commented:
Thanks for your code review. We appreciate it and would address ASAP. ☺

On Thu, Mar 14, 2019, 7:42 AM Alvaro Sanchez Diaz <notifications@github.com>
wrote:

> *@heyapricot* requested changes on this pull request.
> Overkill 💣
>
> Hey! Congratulations on building your first big ruby project! The game
> works and you even added some tests, which is very nice. However, there are
> elements that I think are overkill. They could be simplified. I've left
> some comments. Please address them and after send a new code review request.
>
>    - Additional observation: I like the idea of using a module but it's
>    not implemented appropiately. Board is the only class inside the
>    module. Either avoid the use of the module for now or group all the classes
>    into a single module
>
> As always feel free to reach me on crisp for any kind of help/support
> request. Thanks
> ------------------------------
>
> In lib/classes/game.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265375176>:
>
> > +
>
> +		def get_move(move = gets.chomp)
>
> +			keyboard_to_position(move)
>
> +		end
>
> +
>
> +		def game_over_mssg
>
> +			return "#{@current_player.name} won!" if board.game_over == :winner
>
> +			return 'The game ended in a tie' if board.game_over == :draw
>
> +		end
>
> +
>
> +		def play
>
> +			puts "#{@current_player.name} is the first player!"
>
> +			while true
>
> +				board.formatted_grid
>
> +    		puts ''
>
> +    		puts solicit_move
>
>
>
>    - This could be a matter of style, but I don't like the puts ' '
>    - For solicit_move can we put the print inside the method?
>    - Same with board.formatted_grid, can we wrap it around a method?
>    print_board or something similar
>
> ------------------------------
>
> In lib/classes/game.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265377182>:
>
> > +		private
>
> +
>
> +		def keyboard_to_position(move)
>
> +			keyboard_map = {
>
> +				'1' => [0, 0],
>
> +				'2' => [1, 0],
>
> +				'3' => [2, 0],
>
> +				'4' => [0, 1],
>
> +				'5' => [1, 1],
>
> +				'6' => [2, 1],
>
> +				'7' => [0, 2],
>
> +				'8' => [1, 2],
>
> +				'9' => [2, 2]
>
> +			}
>
> +			keyboard_map[move]
>
> +		end
>
>
> Cool idea 😎
> ------------------------------
>
> In lib/classes/game.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265378136>:
>
> > +		end
>
> +
>
> +		def game_over_mssg
>
> +			return "#{@current_player.name} won!" if board.game_over == :winner
>
> +			return 'The game ended in a tie' if board.game_over == :draw
>
> +		end
>
> +
>
> +		def play
>
> +			puts "#{@current_player.name} is the first player!"
>
> +			while true
>
> +				board.formatted_grid
>
> +    		puts ''
>
> +    		puts solicit_move
>
> +				col, row = get_move
>
> +				puts ''
>
> +    		move = board.set_cell(row, col, @current_player.turn)
>
>
> Why is the player storing the current turn? If we have two players that
> means we are storing the turn twice, right? There should be a better way...
> ------------------------------
>
> In lib/classes/board.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265378926>:
>
> > @@ -0,0 +1,77 @@
>
> +module OurTicTacToe
>
> +	class Board
>
> +		attr_reader :grid
>
> +		def initialize(grid_board = {})
>
> +			@grid = grid_board.fetch(:grid, default_board)
>
> +		end
>
> +
>
> +		def get_cell(row, col)
>
> +			grid[row][col]
>
> +		end
>
> +
>
> +		def set_cell(row, col, value)
>
> +			cell = get_cell(row, col)
>
> +			cell.value.empty? ? cell.value = value : nil
>
>
> I like the ||=
> <http://www.rubyinside.com/what-rubys-double-pipe-or-equals-really-does-5488.html>
> operator for this kind of thing
> ------------------------------
>
> In lib/classes/board.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265379737>:
>
> > @@ -0,0 +1,77 @@
>
> +module OurTicTacToe
>
> +	class Board
>
> +		attr_reader :grid
>
> +		def initialize(grid_board = {})
>
> +			@grid = grid_board.fetch(:grid, default_board)
>
>
> What's the point on wrapping the board array in a hash? In the end you are
> still storing the moves in an array (Which is inside a hash)
> ------------------------------
>
> In lib/classes/cell.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265380830>:
>
> > @@ -0,0 +1,8 @@
>
> +module OurTicTacToe
>
> +	class Cell
>
> +		attr_accessor :value
>
> +		def initialize(value = '')
>
> +			@value = value
>
> +		end
>
> +	end
>
>
> 🤔 If we are only using the Cell class to store a value, do we really
> need a class?
> ------------------------------
>
> In lib/classes/game.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265514116>:
>
> > +
>
> +		def play
>
> +			puts "#{@current_player.name} is the first player!"
>
> +			while true
>
> +				board.formatted_grid
>
> +    		puts ''
>
> +    		puts solicit_move
>
> +				col, row = get_move
>
> +				puts ''
>
> +    		move = board.set_cell(row, col, @current_player.turn)
>
> +    		while move.nil?
>
> +    			puts solicit_move + ' Please enter a number in an empty cell.'
>
> +					col, row = get_move
>
> +					puts ''
>
> +    			move = board.set_cell(row, col, @current_player.turn)
>
> +    		end	
>
>
> Please watch the indentation of your files.
> We can move the while block inside a method. It would look nicer
> ------------------------------
>
> In lib/classes/game.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265515693>:
>
> > +				col, row = get_move
>
> +				puts ''
>
> +    		move = board.set_cell(row, col, @current_player.turn)
>
> +    		while move.nil?
>
> +    			puts solicit_move + ' Please enter a number in an empty cell.'
>
> +					col, row = get_move
>
> +					puts ''
>
> +    			move = board.set_cell(row, col, @current_player.turn)
>
> +    		end	
>
> +				if @board.game_over
>
> +					puts game_over_mssg
>
> +					board.formatted_grid
>
> +					return
>
> +				else
>
> +					switch_players
>
> +				end
>
>
> If we move switch_players to another part of the while loop we can
> rewrite the previous block to something like:
> break if @board.game_over
> We can then call game_over_mssg out of the while loop. What do you think?
> ------------------------------
>
> In lib/classes/board.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265518649>:
>
> > +
>
> +			false
>
> +		end
>
> +
>
> +		def formatted_grid
>
> +			grid.each_with_index do |row, outer|
>
> +				r = row.each_with_index.map do |col, inner|
>
> +					if col.value.empty?
>
> +						case outer
>
> +						when 0
>
> +							inner + 1
>
> +						when 1
>
> +							inner + 4
>
> +						when 2
>
> +							inner + 7
>
> +						end
>
>
> If we use math, we can get rid of the case block.
> The three operations are very similar: inner + number
> Which statement would allow us to get 1,4,7? Use the elements you have:
> outer, inner
> ------------------------------
>
> In lib/classes/game.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265519553>:
>
> > +		private
>
> +
>
> +		def keyboard_to_position(move)
>
> +			keyboard_map = {
>
> +				'1' => [0, 0],
>
> +				'2' => [1, 0],
>
> +				'3' => [2, 0],
>
> +				'4' => [0, 1],
>
> +				'5' => [1, 1],
>
> +				'6' => [2, 1],
>
> +				'7' => [0, 2],
>
> +				'8' => [1, 2],
>
> +				'9' => [2, 2]
>
> +			}
>
> +			keyboard_map[move]
>
> +		end
>
>
> Maybe I'm just lazy, but this seems like a lot of writing (Also, it makes
> us vulnerable to a typo). What if we had a method that populated the hash?
> The hash keys are just a sequence from (1..9). Can we use math to
> generate the array to which the index maps? 🤔🧮
> ------------------------------
>
> In lib/classes/board.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265521180>:
>
> > +						when 1
>
> +							inner + 4
>
> +						when 2
>
> +							inner + 7
>
> +						end
>
> +					else
>
> +						col.value
>
> +					end
>
> +				end
>
> +				puts r.join(' ')
>
> +			end
>
> +		end
>
> +
>
> +		def winning_positions
>
> +			grid + grid.transpose + diagonals
>
> +		end
>
>
> Not bad *per se* but we are generating a whole new matrix by calling
> grid.transpose just to make the checks. You can leave it like this, but I
> want you to think about performance when building apps. Do we really need
> to generate a whole new matrix just to check for the columns?
>
> Life is full of tradeoffs. In programming it may come in the form of
> commodity vs performance.
> ------------------------------
>
> In lib/classes/board.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265521370>:
>
> > +					end
>
> +				end
>
> +				puts r.join(' ')
>
> +			end
>
> +		end
>
> +
>
> +		def winning_positions
>
> +			grid + grid.transpose + diagonals
>
> +		end
>
> +
>
> +		def diagonals
>
> +			[
>
> +				[get_cell(0, 0), get_cell(1, 1), get_cell(2, 2)],
>
> +				[get_cell(0, 2), get_cell(1, 1), get_cell(2, 0)]
>
> +			]
>
> +		end
>
>
> I try to avoid hardcoded values when possible. Can we generate the
> combinations for the diagonals?
> ------------------------------
>
> In lib/classes/core_extensions.rb
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#discussion_r265527379>:
>
> > +	def all_empty?
>
> +		self.all? { |item| item.to_s.empty? }
>
> +	end
>
> +
>
> +	def all_same?
>
> +		self.all? { |item| item == self[0] }
>
> +	end
>
> +
>
> +	def any_empty?
>
> +		self.any? { |item| item.to_s.empty? }
>
> +	end
>
> +
>
> +	def none_empty?
>
> +		!any_empty?
>
> +	end
>
> +end
>
>
> This is a clever idea but I'm not crazy about the extension of core
> elements. This discussion enters the realm of personal
> preference/philosophy. This is the best info I could find about it
> <https://stackoverflow.com/questions/16504369/is-it-proper-ruby-style-to-extend-built-in-classes>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/adriaanbd/tic-tac-toe/pull/7#pullrequestreview-214265119>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Adv63ywRsDA1AZsDUPLkYi9M--VNrR-Bks5vWjWygaJpZM4bwCeI>
> .
>

mcrd25 commented:
Re-Reading the review! loving it! Thanks for getting back to us so quickly again!  @heyapricot

Commit:
This closes issue #8

Commit:
Updated test for that matches changes made fixing issue #8

Commit:
This fixes #10

Commit:
Changed turn to avatar where necessary (ref iss 10) and closes #11

Commit:
Fixes indentation issue

Commit:
try to fix indentation in game.rb

Commit:
This closes #12

Commit:
Got rid of case block in formatted_grid method and simplified to single statement

Commit:
This closes #14

mcrd25 commented:
Hi, @heyapricot Board class is not only one inside OurTicTacToe Module - the other classes are also as well.

Commit:
Merge pull request #16 from adriaanbd/code-review

Code review

heyapricot commented on lib/classes/board.rb:
```diff
@@ -24,18 +24,7 @@ def game_over
-					else
-						col.value
-					end
+					col.value.empty? ? inner +  1 + (3 * outer) : col.value
```
Can we store `row.length` in a variable and use it instead of the `3` to avoid [magic numbers?](https://www.quora.com/What-is-a-magic-number-in-programming)

mcrd25 replied:
Thanks will do that now. Resolved with commit  [acf4f8bba97343d448e80321113fee525ea50369](https://github.com/adriaanbd/tic-tac-toe/pull/18/commits/acf4f8bba97343d448e80321113fee525ea50369)

heyapricot commented on lib/classes/game.rb:
```diff
@@ -12,59 +12,66 @@ def switch_players
+      end
+      print_board
+      game_over_mssg
+    end
```
This loop looks so good ❤️

heyapricot commented on lib/classes/game.rb:
```diff
@@ -12,59 +12,66 @@ def switch_players
+        solicit_move
+        col, row = get_move
+        move = board.set_cell(row, col, @current_player.avatar)
+        solicit_until(move)
```
I'm curious. Why you use `solicit_until` here and before you use `print_board`, `solicit_move` and `col, row = get_move`?

mcrd25 replied:
Because we only want to output a different message if they initially placed move in space that is already occupied, so we want the initial then if they move in empty space then we let them know. 
However, we can simplify.

mcrd25 replied:
Resolved in commit  [acf4f8bba97343d448e80321113fee525ea50369](https://github.com/adriaanbd/tic-tac-toe/pull/18/commits/acf4f8bba97343d448e80321113fee525ea50369)

heyapricot commented on lib/classes/game.rb:
```diff
@@ -12,59 +12,66 @@ def switch_players
+				puts "#{@current_player.name}: Enter a number between 1 and 9.\n"
+			else
+				puts "#{@current_player.name}: Enter a number between 1 and 9. Please enter a number for an empty cell."
+			end
```
If you have an `else` statement use `if`

mcrd25 replied:
Thanks, resolved with commit [acf4f8bba97343d448e80321113fee525ea50369](https://github.com/adriaanbd/tic-tac-toe/pull/18/commits/acf4f8bba97343d448e80321113fee525ea50369)

heyapricot reviewed (changes requested):
# The main loop looks great 🌀
👋🏼 Hello! You got the core working very well. I've added additional observations and replied to your comments.
You still haven't addressed my concerns about the use of `Module`. 

Please address the comments and submit a new code review request. Feel free to reach out over crisp for any technical questions. Thanks!

Commit:
Removed unless statement, simplified soliciting of move to fewer blocks and changed displayed value calculation to utilize row.length rather than 3

adriaanbd committed:
Updated cell value to nil and others

Commit:
Removed use of hash in board class init. Removed use of OurTicTacToe Module and reverted back to original way of doing set_cell due to unforseen breaking errors

Commit:
Merge pull request #18 from adriaanbd/review2

Review2

mcrd25 commented:
> # The main loop looks great 🌀
> 👋🏼 Hello! You got the core working very well. I've added additional observations and replied to your comments.
> You still haven't addressed my concerns about the use of `Module`.
> 
> Please address the comments and submit a new code review request. Feel free to reach out over crisp for any technical questions. Thanks!

Hi @heyapricot  we addressed the module stuff and removed it from use in this project as it really was not necessary as you said.

We were using it wrong, and we really appreciate your feedback :)

See commit [a90931c9cbcf66e3b33aa777b90d3557fee8fbf1](https://github.com/adriaanbd/tic-tac-toe/pull/18/commits/a90931c9cbcf66e3b33aa777b90d3557fee8fbf1)

Commit:
Update README.md

heyapricot reviewed (approved):
# The project structure is rock solid 🗿
I really like the way the project is structured now. I left some comments for improvements.
Nexts steps are:
- Address the comments I've left
- Solve the remaining issues (Particularly the use of `.transpose`)

And that's it 🚀

This is a milestone for your project and you did a great job. Well done 🤩
